apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/DLQGrpcRdaSource.java

```diff
@@ -230,60 +230,62 @@ Processor dlqProcessingLogic(
 
-              if (processed > 0
-                  && dao.updateState(startingSequenceNumber, type, MessageError.Status.RESOLVED)
-                      > 0) {
-                log.info(
-                    "{} claim with sequence ({}) processed successfully, marking as resolved",
-                    claimType,
-                    startingSequenceNumber);
-              }
-            } else {
-              // We didn't get the sequence number we wanted, which means it's obsolete
-              if (dao.updateState(startingSequenceNumber, type, MessageError.Status.OBSOLETE) > 0) {
-                log.info(
-                    "{} claim with sequence({}) was not returned, marking as obsolete",
-                    claimType,
-                    startingSequenceNumber);
+                if (processed > 0
+                    && dao.updateState(startingSequenceNumber, type, MessageError.Status.RESOLVED)
+                        > 0) {
+                  log.info(
+                      "{} claim with sequence ({}) processed successfully, marking as resolved",
+                      claimType,
+                      startingSequenceNumber);
+                }
               } else {
-                log.error(
-                    "{} claim with sequence({}) was not returned, but failed to mark obsolete",
-                    claimType,
-                    startingSequenceNumber);
+                // We didn't get the sequence number we wanted, which means it's obsolete
+                if (dao.updateState(startingSequenceNumber, type, MessageError.Status.OBSOLETE)
+                    > 0) {
+                  log.info(
+                      "{} claim with sequence({}) was not returned, marking as obsolete",
+                      claimType,
+                      startingSequenceNumber);
+                } else {
+                  log.error(
+                      "{} claim with sequence({}) was not returned, but failed to mark obsolete",
+                      claimType,
+                      startingSequenceNumber);
+                }
               }
             }
+            responseStream.cancelStream("No further messages need to be ingested.");
+          } catch (GrpcResponseStream.StreamInterruptedException ex) {
+            // If our thread is interrupted we cancel the stream so the server knows we're done
+            // and then shut down normally.
+            responseStream.cancelStream("shutting down due to InterruptedException");
+            processResult.setInterrupted(true);
+          } catch (Exception e) {
+            responseStream.cancelStream("shutting down due to Exception");
+            // If we failed to process the claim, it stays in the DLQ, nothing to do.
+            log.error(
+                "Failed to process "
+                    + claimType
+                    + " message with sequence: "
+                    + startingSequenceNumber,
+                e);
```

Comment:
Is there any chance `cancelStream()` can throw?  In which case should we `setInterrupted()` / `log.error()` first before canceling the stream to ensure they happen?

Reply:
Good catch.  I'll change that.